x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/TrainedModelStatsService.java

```diff
@@ -0,0 +1,188 @@
+        "     && ctx._source.total_time_spent_millis >= params.total_time_spent_millis\n" +
+        "     && ctx._source.failure_count >= params.failure_count) {\n" +
+        "      ctx.op = 'none';\n" +
+        "      return;\n" +
+        "    }\n" +
+        "    ctx._source.missing_all_fields_count = params.missing_all_fields_count;\n" +
+        "    ctx._source.inference_count = params.inference_count;\n" +
+        "    ctx._source.total_time_spent_millis = params.total_time_spent_millis;\n" +
+        "    ctx._source.failure_count = params.failure_count;\n" +
+        "    ctx._source.time_stamp = params.time_stamp;";
+    private static final ToXContent.Params FOR_INTERNAL_STORAGE_PARAMS =
+        new ToXContent.MapParams(Collections.singletonMap(ToXContentParams.FOR_INTERNAL_STORAGE, "true"));
+
+    private final Map<String, InferenceStats> statsQueue;
+    private final ResultsPersisterService resultsPersisterService;
+    private final OriginSettingClient client;
+    private final IndexNameExpressionResolver indexNameExpressionResolver;
+    private final ThreadPool threadPool;
+    private volatile Scheduler.Cancellable scheduledFuture;
+    private volatile boolean verifiedStatsIndexCreated;
+    private volatile boolean stopped;
+    private volatile ClusterState clusterState;
+
+    public TrainedModelStatsService(ResultsPersisterService resultsPersisterService,
+                                    OriginSettingClient client,
+                                    IndexNameExpressionResolver indexNameExpressionResolver,
+                                    ClusterService clusterService,
+                                    ThreadPool threadPool) {
+        this.resultsPersisterService = resultsPersisterService;
+        this.client = client;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        this.threadPool = threadPool;
+        this.statsQueue = new ConcurrentHashMap<>();
+
+        clusterService.addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void beforeStart() {
+                start();
+            }
+
+            @Override
+            public void beforeStop() {
+                stop();
+            }
+        });
+        clusterService.addListener((event) -> this.clusterState = event.state());
+    }
+
+    public void queueStats(InferenceStats stats) {
+        statsQueue.put(InferenceStats.docId(stats.getModelId(), stats.getNodeId()), stats);
+    }
+
+    void stop() {
+        stopped = true;
+        statsQueue.clear();
+
+        ThreadPool.Cancellable cancellable = this.scheduledFuture;
+        if (cancellable != null) {
+            cancellable.cancel();
+        }
+    }
+
+    void start() {
+        stopped = false;
+        scheduledFuture = threadPool.scheduleWithFixedDelay(this::persistStats,
+            TimeValue.timeValueSeconds(1),
+            MachineLearning.UTILITY_THREAD_POOL_NAME);
+    }
+
+    void persistStats() {
```

Comment:
```suggestion
    void updateStats() {
```

To me `persistStats()` makes me think they are being overwritten and update is clearer. That's how I see it anyway many people may disagree